src/charm.py

```diff
@@ -0,0 +1,257 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical
+# See LICENSE file for licensing details.
+
+"""Charm the calibre-web application."""
+
+import logging
+import typing
+from pathlib import Path
+from typing import cast
+
+import ops
+
+CONTAINER_NAME = 'calibre-web'
+SERVICE_NAME = 'calibre-web'
+STORAGE_NAME = 'books'
+LIBRARY_WRITE_ACTION = 'library-write'
+LIBRARY_WRITE_CONFIG = 'library-write'
+LIBRARY_INFO_ACTION = 'library-info'
+LIBRARY_INFO_FORMAT_PARAM = 'format'
+
+LibraryInfoFormat: typing.TypeAlias = typing.Literal['tree', 'ls-1']
+LIBRARY_INFO_FORMATS = typing.get_args(LibraryInfoFormat)
+
+LibraryWriteBehaviour: typing.TypeAlias = typing.Literal['skip', 'clean']
+LIBRARY_WRITE_BEHAVIOURS = typing.get_args(LibraryWriteBehaviour)
+
+
+CalibreWebLayerDict = typing.TypedDict(
+    'CalibreWebLayerDict',
+    {
+        'summary': str,
+        'description': str,
+        'services': dict[str, ops.pebble.ServiceDict],
+    },
+    total=True,
+)
+
+
+class CalibreWebCharm(ops.CharmBase):
+    """Charm the calibre-web application."""
+
+    def __init__(self, framework: ops.Framework) -> None:
+        super().__init__(framework)
+        framework.observe(self.on.collect_unit_status, self._on_collect_status)
+        framework.observe(self.on[CONTAINER_NAME].pebble_ready, self._on_pebble_ready)
+        framework.observe(self.on[STORAGE_NAME].storage_attached, self._on_storage_attached)
+        framework.observe(self.on.config_changed, self._on_config_changed)
+        framework.observe(self.on[LIBRARY_WRITE_ACTION].action, self._on_library_write)
+        framework.observe(self.on[LIBRARY_INFO_ACTION].action, self._on_library_info)
+        self.unit.set_ports(8083)
+
+    def _on_collect_status(self, event: ops.CollectStatusEvent) -> None:
+        logger.debug("_on_collect_status")
+        event.add_status(ops.ActiveStatus())
+        _, status = self._get_library_write_behaviour()
+        event.add_status(status)
+
+    def _on_pebble_ready(self, event: ops.PebbleReadyEvent) -> None:
+        """Define and start the workload using the Pebble API."""
+        container = event.workload
+        container.add_layer(SERVICE_NAME, {**self.get_pebble_layer()}, combine=True)
```

Comment:
This is just to make the type checker happy, but it's not obvious at all that that's why it's done, is it?

My `get_pebble_layer` returns a `TypedDict` with the guaranteed keys that I provide. When unpacking like this, the type checker is happy to convert that into a `ops.pebble.LayerDict`, which has all optional keys (a superset of the keys I provide).

This is safer than a `cast`, since it will catch incompatibilities introduced by any changes to the `ops.pebble.LayerDict` definition, or errors introduced in `get_pebble_layer`, which a `cast` or `type: ignore` would mask.